homeassistant/components/shelly/number.py

```diff
@@ -0,0 +1,116 @@
+
+from .const import AIOSHELLY_DEVICE_TIMEOUT_SEC, CONF_SLEEP_PERIOD, NUMBER_ENTITIES
+from .entity import (
+    BlockAttributeDescription,
+    ShellySleepingBlockAttributeEntity,
+    async_setup_entry_attribute_entities,
+)
+from .utils import get_device_entry_gen
+
+_LOGGER: Final = logging.getLogger(__name__)
+
+SENSORS: Final = {
+    ("device", "valvePos"): BlockAttributeDescription(
+        icon="mdi:pipe-valve",
+        name="Valve Position",
+        unit=PERCENTAGE,
+        available=lambda block: cast(int, block.valveError) != 1,
+        entity_category=EntityCategory.CONFIG,
+    ),
+}
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up sensors for device."""
+    if get_device_entry_gen(config_entry) == 2:
+        return
+
+    if config_entry.data[CONF_SLEEP_PERIOD]:
+        await async_setup_entry_attribute_entities(
+            hass, config_entry, async_add_entities, SENSORS, BlockSleepingNumber
+        )
+
+
+class BlockSleepingNumber(ShellySleepingBlockAttributeEntity, NumberEntity):
+    """Represent a block sleeping number."""
+
+    @property
+    def value(self) -> float:
+        """Return value of number."""
+        if self.block is not None:
+            return cast(float, self.attribute_value)
+
+        return cast(float, self.last_state)
+
+    @property
+    def unit_of_measurement(self) -> str | None:
+        """Return unit of number."""
+        return cast(str, self._unit)
+
+    @property
+    def min_value(self) -> float:
+        """Return minimum value."""
+        return cast(float, NUMBER_ENTITIES[self.wrapper.model]["min"])
+
+    @property
+    def max_value(self) -> float:
+        """Return maximum value."""
+        return cast(float, NUMBER_ENTITIES[self.wrapper.model]["max"])
+
+    @property
+    def step(self) -> float:
+        """Return step increment/decrement value."""
+        return cast(float, NUMBER_ENTITIES[self.wrapper.model]["step"])
+
+    @property
+    def mode(self) -> NumberMode:
+        """Return mode."""
+        return cast(NumberMode, NUMBER_ENTITIES[self.wrapper.model]["mode"])
+
+    async def async_set_value(self, value: float) -> None:
+        """Set value."""
+        # Example for Shelly Valve: http://192.168.188.187/thermostat/0?pos=13.0
+        rest_path = cast(str, NUMBER_ENTITIES[self.wrapper.model]["rest_path"])
+        rest_arg = cast(str, NUMBER_ENTITIES[self.wrapper.model]["rest_arg"])
+        await self._set_state_full_path(rest_path, {rest_arg: value})
+        self.async_write_ha_state()
+
+    async def _set_state_full_path(self, path: str, params: Any) -> Any:
+        """Set block state (HTTP request)."""
+
+        if (
+            cast(bool, NUMBER_ENTITIES[self.wrapper.model]["rest_channel"])
+            and self.block is not None
+        ):
+            path = f"{path}/{self.block.channel}"
```

Comment:
We should find a better way to eliminate this part

Reply:
This is what I asked for in discord ;-)
But as nobody answered I though to start the discussion submitting the PR.

Simone

Reply:
@thecode what about

```python
        blocktype = self.unique_id.split("-")[-2]
        if blocktype != "device" and self.block:
            path = f"{path}/{self.block.channel}"
```

?

Simone

Reply:
New approach fixes this code block